### PR TITLE
Add AC ip address whitelist to domain server

### PR DESCRIPTION
**Pull request by huffman**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 969cbf4cb0160e4ee2e1b36cb8d3e513826d29df available at: https://gist.github.com/b7e86cf99055833a4ee7b1fcb8348b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-969cbf4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-969cbf4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bd1bcaf1a426733c039c892fe16d60f3eb7f9ef5 available at: https://gist.github.com/b7e86cf99055833a4ee7b1fcb8348b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-bd1bcaf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-bd1bcaf.dmg

**Comment by huffman** [comment]
What this does: adds a whitelist where you can specify specific ips or subnets of ACs that you want to allow to connect to your domain server. This does not reject interface clients, only assignment clients. ACs running on your local system should be able to connect by default (`127.0.0.1`).

The list can be found in domain settings in advanced settings, under "Security"

Example entries:

`192.168.0.100` - allows connections from `192.168.0.100` but no other ips
`192.168.0.0/24` - allows connections from any ip in the range`192.168.0.0` - `192.168.0.255`
`39.29.182.256` - invalid ip, this will show up in the logs as an invalid ip, but no error will be shown on the settings page

This might help constructing ranges for testing: http://www.ipaddressguide.com/cidr

What needs to be tested:
 1. When running the sandbox everything connects as expected - you can connect with interface and all of the servers should be available
 1. Connecting an ac from another computer on your network 
   1. You can do this by running the ac from the command line by navigating to the PR install directory and running: `assignment-client.exe -a IP_ADDRESS_OF_DOMAIN_SERVER`
   1. You should not be able to connect, and you should see logs on the DS about rejecting the assignment request
 1. Adding the IP of your other computer to the whitelist should allow you to connect
 1. Ping me on slack and I can help by trying to connect my AC to your DS

**Comment by hifi-gustavo** [comment]
Build console output for commit: d522d03bd252316a9daac6590d3b0d08d0b4f236 available at: https://gist.github.com/b7e86cf99055833a4ee7b1fcb8348b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-d522d03.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-d522d03.dmg

**Comment by OmegaHeron** [comment]
Testing (be a bit - have to build/deploy for my linux remotes)

**Comment by OmegaHeron** [comment]
Testing Windows 10 - basic localhost config.  **Pass**
DS running on Ubuntu Linux - Basic setup with all AC on same machine as DS.  **Pass**

DS and Remote ACs

What I saw was this - machine one running DS only.  Machine two offering to run tasks.  Denied connection to ACs (correctly).  Then I added a specific IP 192.168.0.6 and applied settings.  AC on 192.168.0.6 still could not connect.  Terminated DS and restarted and remote AC was able to connect.

Next;  I see my ACs on second machine in DS list - I'm able to connect to DS on machine one and have services in Interface.  That's all good.  But I see this in DS log every 5 seconds;

`[11/11 15:38:46] [SUPPRESS] [] 4 repeated log entries matching "Received connection from whitelisted ip: [^ ]+ , matches subnet mask: [^ ]+" - Last entry: "Received connection from whitelisted ip: "192.168.0.6" , matches subnet mask: "192.168.0.6/32""`
`[11/11 15:38:46] [DEBUG] [default] Received connection from whitelisted ip: "192.168.0.6" , matches subnet mask: "192.168.0.6/32"`

Also confirmed allowing an IP range works.

Investigate why I had to terminate DS and start again to have white-list setting apply.  I'm running DS on Linux here, but, it seems to work same on Windows.  If I add 192.168.0.100 to my DS on Windows at 192.168.0.6 and white-list 192.168.0.100 then I can't connect AC until I quit DS and start again - applying change via web control panel is not getting it done.

**Comment by huffman** [comment]
@OmegaHeron Let me see if I can fix the settings not applying on "save" bug

File: domain-server/src/DomainServer.cpp [hunk 1]
```diff
@@ -158,6 +158,42 @@ DomainServer::DomainServer(int argc, char* argv[]) :
 
 
     qDebug() << "domain-server is running";
+    static const QString AC_SUBNET_WHITELIST_SETTING_PATH = "security.ac_subnet_whitelist";
+
+    static const Subnet LOCALHOST { QHostAddress("127.0.0.1"), 32 };
+    this->_acSubnetWhitelist = { LOCALHOST };
```
Comment (Atlante45) [inline domain-server/src/DomainServer.cpp]: Why the `this->`?

Comment (huffman) [inline domain-server/src/DomainServer.cpp]: Oops, I was trying to get VS to autocomplete for me

Comment (huffman) [inline domain-server/src/DomainServer.cpp]: Forgot to remove it, doing that now

**Comment by huffman** [comment]
@OmegaHeron Ok, updated, it should now take effect when you save the settings

**Comment by OmegaHeron** [comment]
@huffman - yes - it now applies on my linux DS from web interface save/restart.  👍 Thanks!

**Comment by huffman** [comment]
@OmegaHeron Oh, did you build it yourself locally?

**Comment by OmegaHeron** [comment]
Yup @huffman - I can build AC/DS on Linux in about 2.5 minutes ;)

**Comment by huffman** [comment]
@OmegaHeron Haha ok awesome... I was confused for a moment.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 528b8e93f00c3a6ea5ea4b0627cbf7bdd5d4fe7d available at: https://gist.github.com/b7e86cf99055833a4ee7b1fcb8348b61<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-528b8e9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9053-528b8e9.dmg

**Pull request merged**